File: 15.hs
#!/usr/bin/env stack
-- stack --resolver lts-12.10 script --package containers

{-# OPTIONS_GHC -Wall #-}
-- {-# LANGUAGE TupleSections #-}

import Control.Exception (assert)
import Data.List (intercalate, sort, sortOn)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Maybe (catMaybes, fromJust, isJust, isNothing, listToMaybe)
import Data.Set (Set)
import qualified Data.Set as Set
import Data.Tuple (swap)
import Utils
import Debug.Trace

type Space = Set Coord -- (y, x) for all non-wall coords
data Kind = Elf | Goblin deriving (Eq, Show)
data Unit = Unit { pos_ :: Coord
                 , kind_ :: Kind
                 , hp_ :: Int
                 , ap_ :: Int
                 } deriving Eq
instance Show Unit where
    show (Unit pos Elf hp _) = "E/" ++ show hp ++ "@" ++ show pos
    show (Unit pos Goblin hp _) = "G/" ++ show hp ++ "@" ++ show pos
type Units = Map Coord Unit
data Game = Game { space_ :: Space
                 , units_ :: Units
                 }
type DistanceMap = Map Coord Int

parse :: Int -> String -> Game
parse elfAttack= buildGame . concat . zipWith parseLine [0 ..] . lines where
    parseLine y = zipWith (\x c -> ((y, x), c)) [0 ..]
    buildGame :: [(Coord, Char)] -> Game
    buildGame = go (Set.empty, Map.empty) where
        go (space, units) [] = Game space units
        go (space, units) ((pos, c) : rest) = go (space', units') rest where
            space' = if c /= '#' then Set.insert pos space else space
            units' = case c of
                'E' -> Map.insert pos (Unit pos Elf 200 elfAttack) units
                'G' -> Map.insert pos (Unit pos Goblin 200 3) units
                _ -> units

render :: Game -> String
render (Game space units) =
    unlines $ map renderLine [ymin - 1 .. ymax + 1] where
        ((ymin, xmin), (ymax, xmax)) = boundingBox $ Set.toList space
        renderLine y = map (renderPos y) [xmin - 1 .. xmax + 1] ++ suffix y
        renderPos y x
          | (y, x) `Map.member` units = case units Map.! (y, x) of
                (Unit _ Elf _ _) -> 'E'
                (Unit _ Goblin _ _) -> 'G'
          | (y, x) `Set.member` space = '.'
          | otherwise= '#'
        suffix y = "  " ++ intercalate ", " (
            Map.elems (Map.map show (Map.filter ((== y) . fst . pos_) units)))

around :: Coord -> Space -- coords around the given point, in reading order
around (y, x) = Set.fromDistinctAscList [
                        (pred y, x), (y, pred x), (y, succ x), (succ y, x)]

adjacent :: Space -> Coord -> Space -- adjacent coords that exist in space
adjacent space = Set.intersection space . around

isAdjacent :: Space -> Coord -> Coord -> Bool
isAdjacent space pos1 pos2 = pos2 `Set.member` adjacent space pos1

enemies :: Units -> Unit -> Units
enemies units unit = case kind_ unit of
    Elf -> Map.filter ((== Goblin) . kind_) units
    Goblin -> Map.filter ((== Elf) . kind_) units

hasEnemies :: Units -> Unit -> Bool
hasEnemies units = not. Map.null . enemies units

occupied :: Units -> Coord -> Bool -- Location c is occupied by one of units
occupied units c = c `Set.member` Map.keysSet units

available :: Game -> Space
available (Game space units) = foldr Set.delete space $ Map.keys units

inRange :: Game -> Unit -> Space -- Open squares adjacent to enemies
inRange game@(Game space units) unit = enemyAdjs where
    avail = available game
    enemyPos = Map.keys $ enemies units unit
    enemyAdjs = Set.unions $ map (adjacent avail) enemyPos

bfs :: Space -> DistanceMap -> [(Coord, Int)] -> Space -> DistanceMap
bfs _ graph [] _ = graph
bfs space graph queue seen = bfs space graph' queue'' seen' where
    ((pos, dist) : queue') = queue
    neighbours = adjacent space pos
    unseen = Set.difference neighbours seen
    enqueue = [(c, dist + 1) | c <- Set.toList unseen]
    graph' = Map.union graph $ Map.fromList enqueue
    queue'' = queue' ++ enqueue
    seen' = Set.union seen unseen

distanceFrom :: Space -> Coord -> DistanceMap -- Map space to distance from pos
distanceFrom space pos =
    bfs space (Map.singleton pos 0) [(pos, 0)] (Set.singleton pos)

chooseTarget :: Space -> Space -> Coord -> (Maybe Coord, DistanceMap)
chooseTarget space inrange src =
    let allDistances = Map.fromSet (distanceFrom space) inrange
        reachable = Map.filter (Map.member src) allDistances
        -- find target nearest to src
        (target, distmap) = head $ sortOn nearest (Map.toList reachable)
        nearest (tgt, dmap) = (dmap Map.! src, tgt)
     in if Map.null reachable 
           then trace "  No target chosen" (Nothing, Map.empty)
           else trace ("  Chose target " ++ show target) (Just target, distmap)

findPath :: Space -> Space -> Coord -> Maybe Coord -- Find where to move
findPath space inrange src =
    let (target, distmap) = chooseTarget space inrange src
        -- find adjancent square nearest target
        adjDist = traceShowId $ Map.fromSet (distmap Map.!?) (adjacent space src)
        moveTo = sortOn swap (Map.toList (Map.filter isJust adjDist))
     in if isNothing target || null moveTo
           then Nothing
           else Just (fst $ head moveTo)

enemyInRange :: Game -> Unit -> Maybe Unit
enemyInRange (Game space units) unit = chooseFirst adjacentEnemies where
        enemies' = enemies units unit
        adjacent' = adjacent space $ pos_ unit
        adjacentEnemies = Map.restrictKeys enemies' adjacent'
        chooseFirst = listToMaybe . Map.elems

turn :: Game -> Unit -> Units
turn = phase1

phase1 :: Game -> Unit -> Units -- Are there any enemies?
phase1 (Game space units) unit
    | not $ hasEnemies units unit = units -- no enemies, combat has ended
    | otherwise = units''' where -- lift unit from board while doing its turn
        units' = Map.delete (pos_ unit) units
        (units'', unit') = phase2 (Game space units') unit
        units''' = Map.insert (pos_ unit') unit' units''

phase2 :: Game -> Unit -> (Units, Unit) -- Are there any target squares?
phase2 game@(Game _ units) unit =
    let inrange = inRange game unit
    in if Set.null inrange
        then (units, unit) -- no available target squares, nothing to do
        else phase3 game unit inrange

phase3 :: Game -> Unit -> Space -> (Units, Unit) -- Do we need to move?
phase3 game@(Game space units) unit inrange
    | pos_ unit `Set.member` inrange = phase9 game unit -- already in range
    | otherwise = phase4 game unit inrange -- need to move

phase4 :: Game -> Unit -> Space -> (Units, Unit) -- Figure out where to move
phase4 game@(Game space units) unit inrange =
    let avail = trace ("4. Finding path from " ++ show unit ++ " to " ++ show inrange) $ available game
        moveTo' = findPath avail inrange (pos_ unit)
        moveTo = trace ("  Found: " ++ show moveTo') moveTo'
    in maybe (units, unit) (phase5 game unit) moveTo

phase5 :: Game -> Unit -> Coord -> (Units, Unit) -- move one step
phase5 game@(Game space units) unit pos = trace ("5. Moving " ++  show unit ++ " to " ++ show pos) $ phase6 game unit' where
    avail = available game
    unit' = assert (pos `Set.member` adjacent avail (pos_ unit)) unit { pos_ = pos }

phase6 :: Game -> Unit -> (Units, Unit) -- attack?
phase6 game@(Game space units) unit = (units, unit)

phase9 :: Game -> Unit -> (Units, Unit)
phase9 game unit = (units_ game, unit)

oneRound :: Game -> Game
oneRound (Game space units) =
    Game space (Map.foldl perUnit units units) where
        perUnit units' = turn (Game space units')

initGame :: Int -> IO Game
initGame elfAttack = do
    input <- readFile "15.input"
    return $ parse elfAttack input

smallGame :: IO Game
smallGame = do
    input <- readFile "15.small.input"
    return $ parse 3 input

main :: IO ()
main = do
    -- part 1
    game <- initGame 3
    print '0'
    putStr $ render game
    print '1'
    putStr $ render $ oneRound game

File: 15.py
#!/usr/bin/env python

from io import StringIO
import sys


Debug = False


def debug(s):
    if Debug:
        print(s)


def bounds(coords):
    xmin = min(x for _, x in coords)
    xmax = max(x for _, x in coords)
    ymin = min(y for y, _ in coords)
    ymax = max(y for y, _ in coords)
    return (ymin, xmin), (ymax, xmax)


class Unit:
    def __init__(self, side, pos, attack_power):
        self.side = side
        self.pos = pos
        self.hp = 200
        self.ap = attack_power

    def __str__(self):
        return '{}/{}@{}'.format(self.side, self.hp, self.pos)

    def turn(self, game):
        debug('{}\'s turn:'.format(self))
        if self.hp <= 0:
            debug('  already dead...')
            return False
        enemies = self.enemies(game)
        if not enemies:
            debug('  no enemies')
            return True  # Combat ends
        in_range = self.in_range(game)
        if not in_range:
            debug('  no squares in range')
            return False
        elif self.pos not in in_range:
            try:
                move_to = self.next_move(game, in_range)
            except IndexError:
                debug('  no path to enemies')
                return False
            self.move(move_to)

        if self.pos not in in_range:
            debug('  no enemies in range')
            return False

        suspects = {e for e in enemies if self.pos in game.adjacent(e.pos)}
        debug('  in range of enemies: {}'.format(
            ', '.join(str(s) for s in suspects)))
        victim = sorted(suspects, key=lambda s: (s.hp, s.pos))[0]
        debug('  attacking {}'.format(victim))
        if self.attack(victim):
            debug('  {} has died!'.format(victim))
            game.kill(victim)
        return False

    def enemies(self, game):
        '''Return all remaining enemies in current game.'''
        return {u for u in game.units.values() if u.side != self.side}

    def in_range(self, game):
        '''Return spaces that are in range of any enemy.'''
        return {p for e in self.enemies(game) for p in game.adjacent(e.pos)}

    def next_move(self, game, in_range):
        '''Find the adjacent square that brings us closest to the enemy.'''
        target, d = game.nearest(self.pos, in_range)
        debug('  approaching {} from a distance of {}'.format(target, d))
        neighbour, _ = game.nearest(target, game.adjacent(self.pos))
        debug('  via {}'.format(neighbour))
        return neighbour

    def move(self, dst):
        '''Move to the given location.'''
        self.pos = dst

    def attack(self, victim):
        '''Attack the given unit. Return True iff the victim is killed.'''
        victim.hp -= self.ap
        return victim.hp <= 0


class Game:
    @classmethod
    def parse(cls, path, elf_attack=3, goblin_attack=3):
        space = set()  # Non-wall spaces
        units = {}  # pos -> Unit instance
        with open(path) as f:
            for y, line in enumerate(f):
                for x, c in enumerate(line.strip()):
                    pos = (y, x)
                    if c != '#':
                        space.add(pos)
                        if c == 'E':
                            units[pos] = Unit('E', pos, elf_attack)
                        elif c == 'G':
                            units[pos] = Unit('G', pos, goblin_attack)
        return cls(space, units)

    def __init__(self, space, units):
        self.space = space
        self.units = units

    def render(self):
        (ymin, xmin), (ymax, xmax) = bounds(self.space)
        out = StringIO()
        for y in range(ymin - 1, ymax + 2):
            units = []
            for x in range(xmin - 1, xmax + 2):
                pos = (y, x)
                if pos in self.units:
                    out.write(self.units[pos].side)
                    units.append(self.units[pos])
                elif pos in self.space:
                    out.write('.')
                else:
                    out.write('#')
            out.write('  ')
            out.write(', '.join(str(u) for u in units))
            out.write('\n')
        return out.getvalue()

    def around(self, coord):
        '''Return coords around the given coord, in reading order.'''
        y, x = coord
        return [(y - 1, x), (y, x - 1), (y, x + 1), (y + 1, x)]

    def available(self):
        '''Return spaces that are not currently occupied.'''
        return {pos for pos in self.space if pos not in self.units}

    def adjacent(self, pos, available=None):
        '''Return any available pos adjacent to the given pos.'''
        if available is None:
            available = self.available()
        return {p for p in self.around(pos) if p in available}

    def distance_from(self, pos, available=None):
        '''Return map: p -> dist(p, pos) for p in available.'''
        if available is None:
            available = self.available()
        ret = {pos: 0}  # p -> distance from p to pos
        available.discard(pos)
        adjacents = self.adjacent(pos, available)
        available -= adjacents
        queue = list(adjacents)
        while queue:
            p = queue.pop(0)
            dist = min(ret.get(a, sys.maxsize) for a in self.around(p)) + 1
            ret[p] = dist
            adjacents = self.adjacent(p, available)
            available -= adjacents
            queue.extend(adjacents)
        return ret

    def nearest(self, pos, points, available=None):
        '''Return (p, distance(pos, p)) for the point p nearest to pos.'''
        def dist_then_coord(item):
            p, d = item
            return (d, p)

        d_from_pos = self.distance_from(pos, available)
        point_ds = {p: d for p, d in d_from_pos.items() if p in points}
        return sorted(point_ds.items(), key=dist_then_coord)[0]

    def kill(self, unit):
        del self.units[unit.pos]

    def round(self):
        '''Do one complete round, i.e. give each unit one turn.'''
        for pos, unit in sorted(self.units.items()):
            if unit.hp <= 0:  # already dead
                continue
            del self.units[pos]  # Lift unit from board
            ends = unit.turn(self)
            self.units[unit.pos] = unit  # Replace unit onto board
            if ends:
                return True
        return False

    def until_combat_end(self):
        rounds, end = 0, False
        debug(self.render())
        while not end:
            debug('Round #{}'.format(rounds + 1))
            end = self.round()
            debug(self.render())
            rounds += 1
        return rounds


if __name__ == '__main__':
    # part 1
    game = Game.parse('15.input', elf_attack=3)
    rounds = game.until_combat_end()
    debug(game.render())
    print((rounds - 1) * sum(u.hp for u in game.units.values()))

    # part 2
    for elf_attack in range(4, sys.maxsize):
        debug('Elves attacking with {}'.format(elf_attack))
        game = Game.parse('15.input', elf_attack=elf_attack)
        before = len([u for u in game.units.values() if u.side == 'E'])
        rounds = game.until_combat_end()
        after = len([u for u in game.units.values() if u.side == 'E'])
        debug(game.render())
        if before > after:
            debug('{} elves died in the making of this combat'.format(
                before - after))
        else:
            debug('All elves survived!')
            print((rounds - 1) * sum(u.hp for u in game.units.values()))
            break
